### ISSUE TITLE
Performance Improvement for Graphs

### ISSUE THREAD

**Issue by avose** [avose/GLShell#45]
The new font / text rendering code is actually pretty fast and no longer much of an issue. However, large graphs are still a problem, even ignoring the physics.

Drawing labels on all the nodes of a large graph may, in the worst cases, drop the frame rate by about half, but going from 16fps to 8fps does show that just drawing the node points and the edge lines is slow. If you think about it, adding all that text is adding tens of thousands of textured polygons, yet the performance impact isn't really all that huge considering the extra load on the GPU.

The real problem is the time spent in Python code laying everything out. Python is **_freakishly_** slow (well, I come from programming in C with a focus on high-performance computing (HPC), so my opinion may not be a surprise).

To make things faster, the rendering should be redone to minimize the Python code, and to minimize the number of calls to OpenGL routines. The physics already has numpy arrays with the node positions, so that should be used directly wherever possible.

OpenGL does have support for a good bit of this, even with the old-style, fixed-function API. The following link gives an overview of how to do some of that with `glDrawArrays()` and `glDrawElements()`.
https://eng.libretexts.org/Bookshelves/Computer_Science/Applied_Programming/Book%3A_Introduction_to_Computer_Graphics_(Eck)/03%3A_OpenGL_1.1-_Geometry/3.04%3A_Polygonal_Meshes_and_glDrawArrays

Example code:
```
float coords[6] = { -0.9,-0.9, 0.9,-0.9, 0,0.7 }; // two coords per vertex.
float colors[9] = { 1,0,0, 0,1,0, 1,0,0 }; // three RGB values per vertex.

glVertexPointer( 2, GL_FLOAT, 0, coords ); // Set data type and location.
glColorPointer( 3, GL_FLOAT, 0, colors )

glEnableClientState( GL_VERTEX_ARRAY ); // Enable use of arrays.
glEnableClientState( GL_COLOR_ARRAY );

glDrawArrays( GL_TRIANGLES, 0, 3 ); // Use 3 vertices, starting with vertex 0.
```
http://math.hws.edu/eck/cs424/graphicsbook2018/source/glut/cubes-with-vertex-arrays.c

**Comment by avose** [comment]
For the edges, which would be thought of as indices into the node position array, there is even support for indexing with `glDrawElements()`.

```
float vertexCoords[24] = {  // Coordinates for the vertices of a cube.
    1,1,1,   1,1,-1,   1,-1,-1,   1,-1,1,
    -1,1,1,  -1,1,-1,  -1,-1,-1,  -1,-1,1  };
                        
float vertexColors[24] = {  // An RGB color value for each vertex
    1,1,1,   1,0,0,   1,1,0,   0,1,0,                  
    0,0,1,   1,0,1,   0,0,0,   0,1,1  };
                      
int elementArray[24] = {  // Vertex numbers for the six faces.
    0,1,2,3, 0,3,7,4, 0,4,5,1,                 
    6,2,1,5, 6,5,4,7, 6,7,3,2  };
glVertexPointer( 3, GL_FLOAT, 0, vertexCoords );
glColorPointer( 3, GL_FLOAT, 0, vertexColors );

glEnableClientState( GL_VERTEX_ARRAY );
glEnableClientState( GL_COLOR_ARRAY );

glDrawElements( GL_QUADS, 24, GL_UNSIGNED_INT, elementArray );
```

**Comment by avose** [comment]
The colors can be set into a new array when the graph is created. Then all the rendering code needs to do is to register the existing arrays with OpenGL, and make a couple of calls: `glDrawArrays(GL_POINTS, ...)` for the nodes, and `glDrawElements(GL_LINES, ...)` for the edges.

Numpy can be used to do the projection of all the node positions to screen coordinates, so `gluProject()` doesn't need to be called individually for each of the nodes. However, I don't currently see a way around the need to loop through all the nodes in Python at least once to draw all the text labels.

**Comment by avose** [comment]
Ah, and a more traditional pass may be needed when doing a rendering pass for selection, as I don't see a supported array for names to replace `glLoadName()`...

Also, for the selected nodes and for the search result nodes, which are drawn larger, I would need to make a pass for each size.. Gets slightly complicated, I suppose.. but not too bad probably.

**Comment by avose** [comment]
OK, this has been done for the edges. They were a fairly simple case as they're all the same color right now, and nothing else needs to be done with the coordinates, but draw them.

The nodes are a little more complicated, will get to them a bit later.

**Comment by avose** [comment]
I created a new branch `opengl-performance` for this work. The nodes are now grouped into "kinds", which are each drawn in their own pass. The kinds now include: file, dir, selected, and search result. This allows very fast rendering with different styles for each of the different node kinds.

The selection pass is now a simplified loop which is just drawing unstyled points to get the `glLoadName()` emitted.

The part that's left here is to get the equivalent of `gluProject()` in place, so I have coordinates at which to draw the labels.

**Comment by avose** [comment]
I did a reasonable job with optimizing the label drawing pass. At least now, only nodes with labels will need to do a `gluProject()`, and the code for all this is much more streamlined. All the cascading conditionals to choose a label style are now done with a single dictionary lookup.

I can do the `gluProject()` equivalent at some point in the future if needed, but things here are so vastly improved as to where they were before I started this process, that I think it's fine to close the issue for now.
